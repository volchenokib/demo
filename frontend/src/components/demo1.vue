<template>
  <v-container>
    <v-layout column>
      <v-card class="pa-4 my-0 mx-auto dropDown" max-width="650" elevation="8">
        <v-form v-model="demoForm" ref="form" name="demoForm" class="text-xs-center">
          <v-layout row wrap justify-space-between>
            <!-- picker #1 -->
            <v-flex xs12 sm5>
              <v-menu
                ref="menuFrom"
                v-model="menuFrom"
                :close-on-content-click="false"
                lazy
                transition="scale-transition"
                max-width="290px"
                min-width="290px"
                offset-y
                full-width
                :nudge-right="0"
                :nudge-bottom="30"
              >
                <template v-slot:activator="{ on }">
                  <v-text-field
                    v-model="computedDateFrom"
                    label="от"
                    persistent-hint
                    prepend-icon="event"
                    readonly
                    required
                    :rules="fromDateRules"
                    v-on="on"
                    color="primary"
                  ></v-text-field>
                </template>

                <v-date-picker
                  v-model="dateFrom"
                  min="2014-01-01"
                  :max="dateTo || today"
                  no-title
                  @input="menuFrom = false"
                  scrollable
                  first-day-of-week="1"
                  locale="ru-RU"
                  color="primary"
                ></v-date-picker>
              </v-menu>
            </v-flex>
            <!-- picker #2 -->
            <v-flex xs12 sm5>
              <v-menu
                v-model="menuTo"
                :close-on-content-click="false"
                lazy
                transition="scale-transition"
                max-width="290px"
                min-width="290px"
                offset-y
                full-width
                :nudge-right="0"
                :nudge-bottom="30"
              >
                <template v-slot:activator="{ on }">
                  <v-text-field
                    v-model="computedDateTo"
                    label="до"
                    persistent-hint
                    prepend-icon="event"
                    readonly
                    required
                    :rules="toDateRules"
                    v-on="on"
                    color="primary"
                  ></v-text-field>
                </template>

                <v-date-picker
                  v-model="dateTo"
                  :min="dateFrom || '2014-01-01'"
                  :max="today"
                  no-title
                  @input="menuTo = false"
                  scrollable
                  first-day-of-week="1"
                  locale="ru-RU"
                  color="primary"
                ></v-date-picker>
              </v-menu>
            </v-flex>
          </v-layout>

          <!-- radio buttons -->
          <v-radio-group v-model="purchaseType" :mandatory="true">
            <v-radio
              name="singleProvider"
              label="закупка у единственного поставщика"
              value="singleProvider"
              color="primary"
            ></v-radio>
            <v-radio name="eAuction" label="электронный аукцион" value="eAuction" color="primary"></v-radio>
          </v-radio-group>

          <!-- switcher -->
          <v-switch
            name="localStorage"
            :value="true"
            v-model="localStorage"
            label="Взять из локального хранилища"
            color="primary"
          ></v-switch>

          <v-divider></v-divider>

          <!-- button -->
          <v-btn
            ref="downloadButton"
            color="primary"
            flat
            :disabled="buttonDisable"
            :loading="dataLoading"
            @click="submit"
          >
            сформировать
            <template v-slot:loader>
              <span class="d-flex mt-2">
                <v-icon class="loader mr-1" light>cached</v-icon>Загрузка
              </span>
            </template>
          </v-btn>
        </v-form>
      </v-card>
    </v-layout>
  </v-container>
</template>

<style lang="scss">
.loader {
  animation: loader 1s infinite;
  margin-top: -4px;
}
@keyframes loader {
  from {
    transform: rotate(0);
  }
  to {
    transform: rotate(360deg);
  }
}
</style>

<script>
// TODO:
// + formData
// - remove getters
// - error handler
// - notifications
export default {
  name: "demo1",
  data: () => ({
    demoForm: "",
    fromDateRules: [v => !!v || "Это обязательное поле"],
    toDateRules: [v => !!v || "Это обязательное поле"],
    formattedDateFrom: null,
    dateFrom: null,
    formattedDateTo: null,
    dateTo: null,
    menuFrom: false,
    menuTo: false,
    purchaseType: "singleProvider",
    localStorage: false
  }),

  computed: {
    resetForm() {
      return this.$store.getters.formResetNeed;
    },
    dataLoading() {
      return this.$store.getters.getDataState;
    },
    buttonDisable() {
      return this.$store.getters.getButtonState;
    },
    computedDateFrom: {
      set: function() {
        this.dateFrom;
      },
      get: function() {
        if (this.dateFrom) {
          const [year, month, day] = this.dateFrom.split("-");
          return (this.formattedDateFrom = `${day}.${month}.${year}`);
        }
      }
    },
    computedDateTo: {
      set: function() {
        this.dateTo;
      },
      get: function() {
        if (this.dateTo) {
          const [year, month, day] = this.dateTo.split("-");
          return (this.formattedDateTo = `${day}.${month}.${year}`);
        }
      }
    },
    today() {
      var today = new Date();
      var dd = String(today.getDate()).padStart(2, "0");
      var mm = String(today.getMonth() + 1).padStart(2, "0"); //January is 0!
      var yyyy = today.getFullYear();

      return (today = yyyy + "-" + mm + "-" + dd);
    }
  },

  watch: {
    resetForm() {
      this.$refs.form.reset();
      this.dateFrom = null;
      this.dateTo = null;
      this.purchaseType = "singleProvider";
      this.localStorage = false;
      // this.$store.state.form.isReset = false;
      this.$store.commit("API_DATA_SUCCES", false);
    }
  },

  methods: {
    submit() {
      if (this.$refs.form.validate()) {
        const payload = new FormData(demoForm);

        payload.append("dateFrom", this.formattedDateFrom);
        payload.append("dateTo", this.formattedDateTo);

        // Display the keys
        // for (let key of payload.keys()) {
        //   console.log(key);
        // }

        // Display the values
        // for (let value of payload.values()) {
        //   console.log(value);
        // }

        this.$store
          .dispatch("getFile", payload)
          .then(response => this.download(response));
      }
    },

    download(response) {
      const url = window.URL.createObjectURL(new Blob([response.data]));
      const link = document.createElement("a");
      link.href = url;
      link.setAttribute("download", "file.xlsx");
      document.body.appendChild(link);
      link.click();
    }
  }
};
</script>
