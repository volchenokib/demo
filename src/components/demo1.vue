<template>
  <v-container>
    <v-layout column>
      <v-card class="pa-4 my-0 mx-auto dropDown" max-width="650" elevation="8">
        <v-form v-model="demoForm" ref="form" class="text-xs-center">
          <v-layout row wrap justify-space-between>
            <!-- picker #1 -->
            <v-flex xs12 sm5>
              <v-menu
                ref="menuFrom"
                v-model="menuFrom"
                :close-on-content-click="false"
                lazy
                transition="scale-transition"
                max-width="290px"
                min-width="290px"
                offset-y
                full-width
                :nudge-right="0"
                :nudge-bottom="30"
              >
                <template v-slot:activator="{ on }">
                  <v-text-field
                    v-model="computedDateFrom"
                    label="от"
                    persistent-hint
                    prepend-icon="event"
                    readonly
                    required
                    :rules="fromDateRules"
                    v-on="on"
                    color="primary"
                  ></v-text-field>
                </template>

                <v-date-picker
                  v-model="dateFrom"
                  min="2014-01-01"
                  :max="dateTo || today"
                  no-title
                  @input="menuFrom = false"
                  scrollable
                  first-day-of-week="1"
                  locale="ru-RU"
                  color="primary"
                ></v-date-picker>
              </v-menu>
            </v-flex>
            <!-- picker #2 -->
            <v-flex xs12 sm5>
              <v-menu
                v-model="menuTo"
                :close-on-content-click="false"
                lazy
                transition="scale-transition"
                max-width="290px"
                min-width="290px"
                offset-y
                full-width
                :nudge-right="0"
                :nudge-bottom="30"
              >
                <template v-slot:activator="{ on }">
                  <v-text-field
                    v-model="computedDateTo"
                    label="до"
                    persistent-hint
                    prepend-icon="event"
                    readonly
                    required
                    :rules="toDateRules"
                    v-on="on"
                    color="primary"
                  ></v-text-field>
                </template>

                <v-date-picker
                  v-model="dateTo"
                  :min="dateFrom || '2014-01-01'"
                  :max="today"
                  no-title
                  @input="menuTo = false"
                  scrollable
                  first-day-of-week="1"
                  locale="ru-RU"
                  color="primary"
                ></v-date-picker>
              </v-menu>
            </v-flex>
          </v-layout>

          <!-- radio buttons -->
          <v-radio-group v-model="purchaseType" :mandatory="true">
            <v-radio
              label="закупка у единственного поставщика"
              value="singleProvider"
              color="primary"
            ></v-radio>
            <v-radio label="электронный аукцион" value="eAuction" color="primary"></v-radio>
          </v-radio-group>

          <!-- switcher -->
          <v-switch v-model="localStorage" label="Взять из локального хранилища" color="primary"></v-switch>

          <v-divider></v-divider>

          <!-- button -->
          <v-btn
            color="primary"
            flat
            :disabled="buttonDisable"
            :loading="dataLoading"
            @click="download"
          >
            сформировать
            <template v-slot:loader>
              <span class="d-flex mt-2">
                <v-icon class="loader mr-1" light>cached</v-icon>Загрузка
              </span>
            </template>
          </v-btn>
        </v-form>
      </v-card>
    </v-layout>
  </v-container>
</template>

<style lang="scss">
.loader {
  animation: loader 1s infinite;
  margin-top: -4px;
}
@keyframes loader {
  from {
    transform: rotate(0);
  }
  to {
    transform: rotate(360deg);
  }
}
</style>


<script>
export default {
  name: "demo1",
  data: () => ({
    demoForm: "",
    fromDateRules: [v => !!v || "Это обязательное поле"],
    toDateRules: [v => !!v || "Это обязательное поле"],
    formattedDateFrom: null,
    dateFrom: null,
    formattedDateTo: null,
    dateTo: null,
    menuFrom: false,
    menuTo: false,
    purchaseType: "singleProvider",
    localStorage: false
  }),

  computed: {
    resetForm() {
      return this.$store.getters.formResetNeed;
    },
    dataLoading() {
      return this.$store.getters.getDataState;
    },
    buttonDisable() {
      return this.$store.getters.getButtonState;
    },
    computedDateFrom: {
      set: function() {
        this.dateFrom;
      },
      get: function() {
        console.log("dateFrom", this.dateFrom);
        if (this.dateFrom) {
          const [year, month, day] = this.dateFrom.split("-");
          return (this.formattedDateFrom = `${day}.${month}.${year}`);
        }
      }
    },
    computedDateTo: {
      set: function() {
        this.dateTo;
      },
      get: function() {
        if (this.dateTo) {
          const [year, month, day] = this.dateTo.split("-");
          return (this.formattedDateTo = `${day}.${month}.${year}`);
        }
      }
    },
    today() {
      var today = new Date();
      var dd = String(today.getDate()).padStart(2, "0");
      var mm = String(today.getMonth() + 1).padStart(2, "0"); //January is 0!
      var yyyy = today.getFullYear();

      return (today = yyyy + "-" + mm + "-" + dd);
    }
  },

  watch: {
    resetForm() {
      this.$refs.form.reset();
      this.dateFrom = null;
      this.dateTo = null;
      this.purchaseType = "singleProvider";
      this.localStorage = false;
      // this.$store.state.form.isReset = false;
      this.$store.commit("API_DATA_SUCCES", false);
      console.log("watcher", this.$store.state.form.isReset);
    }
  },

  methods: {
    download() {
      if (this.$refs.form.validate()) {
        const payload = {
          dateFrom: this.formattedDateFrom,
          dateTo: this.formattedDateTo,
          purchaseType: this.purchaseType,
          localStorage: this.localStorage
        };

        this.$store.dispatch("downloadFile", payload);
        console.log("payload", payload);
      }
    }
  }
};
</script>
