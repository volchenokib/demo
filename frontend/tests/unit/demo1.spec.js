import Vue from 'vue';
import Vuex from 'vuex';
import Vuetify from 'vuetify';
import { mount, createLocalVue } from '@vue/test-utils';
import demo1 from '@/components/demo1.vue';
import moxios from 'moxios';

Vue.use(Vuex);
Vue.use(Vuetify);
const localVue = createLocalVue();
localVue.use(Vuex, Vuetify);

let wrapper;
let store;
let actions;
let mutations;
let commit;

actions = {
	downloadFile: jest.fn(),
	forceFileDownload: jest.fn()
};

let url = '';
let body = {};

jest.mock('axios', () => ({
	post: (_url, _body) => {
		return new Promise(resolve => {
			url = _url;
			body = _body;
			resolve(true);
		});
	}
}));

store = new Vuex.Store({
	state: {},
	actions,
	mutations,
	commit
});

beforeEach(() => {
	wrapper = mount(demo1, {
		sync: false,
		// propsData: {},
		// mocks: {},
		// stubs: {},
		// methods: {},
		store,
		localVue
	});
});

afterEach(() => {
	wrapper.destroy();
});

describe('download file pipeline', () => {
	it('download button exist', () => {
		expect(wrapper.contains({ ref: 'downloadButton' })).toBe(true);
	});
	it('by clicking on the download button, the download method starts', () => {
		// wrapper.setMethods({ download: jest.fn() });
		// wrapper.find({ ref: 'downloadButton' }).trigger('click');
		// expect(wrapper.vm.download).toHaveBeenCalled();
	});

	describe('download method works correctly', () => {
		// const wrapper = mount(demo1, { store, localVue });
		it('form validation started', () => {
			expect(wrapper.contains('button')).toBe(true);
		});
		it('if the form validation passed, downLoadFile method was dispatched with correct payload', () => {
			expect(wrapper.contains('button')).toBe(true);
		});

		// it('text is correct', () => {
		// 	const buttonContent = wrapper.find('.v-btn__content');
		// 	expect(buttonContent.text()).toMatch('сформировать');
		// });
		// it('runs action correctly', () => {
		// 	const textInput = wrapper.findAll('input[type="text"]');
		// 	textInput.setValue('01-01-2018');
		// 	wrapper.find('button').trigger('click');
		// 	expect(actions.downloadFile).toHaveBeenCalled();
		// 	expect(actions.downloadFile).toHaveBeenCalledTimes(1);
		// 	// expect(actions.downloadFile).toHaveBeenLastCalledWith(store, {
		// 	// 	dateFrom: '01.01.2018',
		// 	// 	dateTo: '01.01.2018',
		// 	// 	purchaseType: 'singleProvider',
		// 	// 	localStorage: false
		// 	// });
		// });
	});

	describe('downloadFile action works correctly', () => {
		it('correct API endpoint was used', async () => {
			await actions.downloadFile();

			expect(url).toBe('https://78.media.tumblr.com/tumblr_m39nv7PcCU1r326q7o1_500.png');
		});
		it('payload is correct', () => {});
		it('correct mutation was committed with the result', () => {});
	});

	describe('отправка запроса на сервер', () => {});

	describe('обработка ответа сервера', () => {});
});
