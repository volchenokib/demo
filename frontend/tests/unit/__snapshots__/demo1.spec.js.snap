// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`demo1 component renders correctly 1`] = `
<v-container-stub
  tag="div"
>
  <v-layout-stub
    column=""
    tag="div"
  >
    <v-card-stub
      class="pa-4 my-0 mx-auto dropDown"
      elevation="8"
      maxwidth="650"
      tag="div"
    >
      <v-form-stub
        class="text-xs-center"
        value="true"
      >
        <v-layout-stub
          justify-space-between=""
          row=""
          tag="div"
          wrap=""
        >
          <v-flex-stub
            sm5=""
            tag="div"
            xs12=""
          >
            <v-menu-stub
              closedelay="0"
              closeonclick="true"
              contentclass=""
              fullwidth="true"
              lazy="true"
              maxheight="auto"
              maxwidth="290px"
              minwidth="290px"
              nudgebottom="30"
              nudgeleft="0"
              nudgeright="0"
              nudgetop="0"
              nudgewidth="0"
              offsety="true"
              opendelay="0"
              openonclick="true"
              origin="top left"
              transition="scale-transition"
            >
               
              <v-date-picker-stub
                color="primary"
                eventcolor="warning"
                firstdayofweek="1"
                locale="ru-RU"
                max="2019-04-15"
                min="2014-01-01"
                nexticon="$vuetify.icons.next"
                notitle="true"
                previcon="$vuetify.icons.prev"
                scrollable="true"
                showcurrent="true"
                type="date"
                width="290"
              />
            </v-menu-stub>
          </v-flex-stub>
           
          <v-flex-stub
            sm5=""
            tag="div"
            xs12=""
          >
            <v-menu-stub
              closedelay="0"
              closeonclick="true"
              contentclass=""
              fullwidth="true"
              lazy="true"
              maxheight="auto"
              maxwidth="290px"
              minwidth="290px"
              nudgebottom="30"
              nudgeleft="0"
              nudgeright="0"
              nudgetop="0"
              nudgewidth="0"
              offsety="true"
              opendelay="0"
              openonclick="true"
              origin="top left"
              transition="scale-transition"
            >
               
              <v-date-picker-stub
                color="primary"
                eventcolor="warning"
                firstdayofweek="1"
                locale="ru-RU"
                max="2019-04-15"
                min="2014-01-01"
                nexticon="$vuetify.icons.next"
                notitle="true"
                previcon="$vuetify.icons.prev"
                scrollable="true"
                showcurrent="true"
                type="date"
                width="290"
              />
            </v-menu-stub>
          </v-flex-stub>
        </v-layout-stub>
         
        <v-radio-group-stub
          backgroundcolor=""
          column="true"
          errorcount="1"
          errormessages=""
          height="auto"
          mandatory="true"
          messages=""
          rules=""
          successmessages=""
          value="singleProvider"
          valuecomparator="function deepEqual(a, b) {
    if (a === b) return true;
    if (a instanceof Date && b instanceof Date) {
        // If the values are Date, they were convert to timestamp with getTime and compare it
        if (a.getTime() !== b.getTime()) return false;
    }
    if (a !== Object(a) || b !== Object(b)) {
        // If the values aren't objects, they were already checked for equality
        return false;
    }
    var props = Object.keys(a);
    if (props.length !== Object.keys(b).length) {
        // Different number of props, don't bother to check
        return false;
    }
    return props.every(function (p) {
        return deepEqual(a[p], b[p]);
    });
}"
        >
          <v-radio-stub
            color="primary"
            label="закупка у единственного поставщика"
            officon="$vuetify.icons.radioOff"
            onicon="$vuetify.icons.radioOn"
            ripple="true"
            value="singleProvider"
          />
           
          <v-radio-stub
            color="primary"
            label="электронный аукцион"
            officon="$vuetify.icons.radioOff"
            onicon="$vuetify.icons.radioOn"
            ripple="true"
            value="eAuction"
          />
        </v-radio-group-stub>
         
        <v-switch-stub
          backgroundcolor=""
          color="primary"
          errorcount="1"
          errormessages=""
          label="Взять из локального хранилища"
          messages=""
          ripple="true"
          rules=""
          successmessages=""
          valuecomparator="function deepEqual(a, b) {
    if (a === b) return true;
    if (a instanceof Date && b instanceof Date) {
        // If the values are Date, they were convert to timestamp with getTime and compare it
        if (a.getTime() !== b.getTime()) return false;
    }
    if (a !== Object(a) || b !== Object(b)) {
        // If the values aren't objects, they were already checked for equality
        return false;
    }
    var props = Object.keys(a);
    if (props.length !== Object.keys(b).length) {
        // Different number of props, don't bother to check
        return false;
    }
    return props.every(function (p) {
        return deepEqual(a[p], b[p]);
    });
}"
        />
         
        <v-divider-stub />
         
        <v-btn-stub
          activeclass="v-btn--active"
          color="primary"
          flat="true"
          tag="button"
          type="button"
        >
          
          сформировать
          
        </v-btn-stub>
      </v-form-stub>
    </v-card-stub>
  </v-layout-stub>
</v-container-stub>
`;
